gateway/src/resty/log/log.lua

```diff
@@ -0,0 +1,92 @@
+local errlog = require "ngx.errlog"
+local base = require "resty.core.base"
+local ffi = require 'ffi'
+
+local C = ffi.C
+local FFI_BAD_CONTEXT = base.FFI_BAD_CONTEXT
+local find = string.find
+local getinfo = debug.getinfo
+local reverse = string.reverse
+local select = select
+local sub = string.sub
+local tostring = tostring
+
+local get_request = base.get_request
+
+
+ffi.cdef[[
+int ngx_http_lua_ffi_get_phase(ngx_http_request_t *r, char **err)
+]]
+
+local _M = {}
+
+local function get_prefix_from_info(info)
+  local reverse_str = reverse(info.short_src)
+  local first_match = find(reverse_str, "/")
+  if not first_match  then
+    return info.short_src
+  end
+  local reverse_filename = sub(reverse_str, 0, first_match - 1)
+
+  local result = string.reverse(reverse_filename) .. ":" .. info.currentline
+
+  -- If the function name is present, just append to the prefix.
+  if info.name then
+    result = result .. ": " .. info.name .. "()"
+  end
+  return result
+end
+
+-- is_valid_request returns true if is a request. base.get_request returns a
+-- request on init_worker and timer phases, so we need to validate that a valid
+-- request is made.
+local function is_valid_request()
+  local r = get_request()
+  if not r then
+    return nil
+  end
+
+  local id = C.ngx_http_lua_ffi_req_get_method(r)
+  if id == FFI_BAD_CONTEXT then
+    return nil
+  end
+  return true
+end
+
+local function send_log(level, msg, ...)
+  local suffix = ""
+
+  -- Only append the request_id if it's a request, if not will raise an
+  -- exception.
+  local r = is_valid_request()
```

Comment:
We could use this without assigning it to a var, right? Te name of the var is not informative, but the name of the function is.